Add `@cachedproperty` decorator

Pyprohly opened the pull request:
Features:

- Cleaned up constructors by eliminating a bunch of property-related variable assignments (e.g., `self._mod = None`).
- Made property logic simpler and more self contained.
- Property access is about 3x faster (using timeit over 1000000 trials) due to abolished `None` check.

bboe commented:
This PR is awesome. Great work! The other merged PRs result in some conflicts, so once those are resolved we can merge this in.

Pyprohly commented:
I’ve rebased.

Be mindful of the “`New in version |version|.`” I added to the doc string of `cachedproperty`.

Some sort of sphinx magic can be done to replace `|version|` rather than sed.

bboe commented:
> Some sort of sphinx magic can be done to replace |version| rather than sed.

Awesome! Maybe you can whip something up to auto-set that field? Or retroactively hard code it in?

Pyprohly commented:
Oof, cut it close there 😂

I’m not yet super familiar with how all the documentation related stuff is organised that I can whip something up soon.

I can look into it. There‘s no rush.

Pyprohly commented:
Oof, cut it really close there 😂

I’m not yet super familiar with how all the documentation related stuff is organised that I can whip something up soon.

I can look into it. There‘s no rush.
